### ISSUE TITLE
ARM32 pc as dest register

### ISSUE THREAD

**Issue by JonathanSalwan** [JonathanSalwan/Triton#1076]
Hey @cnheitman,

With `unicorn 1.0.3` we have two issues with ARM32 unit tests. Looks like this `pc` register will end up killing us :)). Can you take a look when you got time?

```
Test project /home/jonathan/Works/Tools/Triton/build
      Start  1: TaintRegister
 1/41 Test  #1: TaintRegister ....................   Passed    0.00 sec
      Start  2: InfoRegister
 2/41 Test  #2: InfoRegister .....................   Passed    0.00 sec
      Start  3: IR
 3/41 Test  #3: IR ...............................   Passed    0.00 sec
      Start  4: Simplification
 4/41 Test  #4: Simplification ...................   Passed    0.00 sec
      Start  5: Constraint
 5/41 Test  #5: Constraint .......................   Passed    0.02 sec
      Start  6: TestAPI
 6/41 Test  #6: TestAPI ..........................   Passed    0.01 sec
      Start  7: UnicornAArch64Semantics
 7/41 Test  #7: UnicornAArch64Semantics ..........   Passed    6.39 sec
      Start  8: UnicornX86Semantics
 8/41 Test  #8: UnicornX86Semantics ..............   Passed    1.70 sec
      Start  9: UnicornARM32Semantics1
 9/41 Test  #9: UnicornARM32Semantics1 ...........   Passed    0.58 sec
      Start 10: UnicornARM32Semantics2
10/41 Test #10: UnicornARM32Semantics2 ...........   Passed    0.52 sec
      Start 11: UnicornARM32Semantics3
11/41 Test #11: UnicornARM32Semantics3 ...........   Passed    0.63 sec
      Start 12: UnicornARM32Semantics4
12/41 Test #12: UnicornARM32Semantics4 ...........***Failed    0.18 sec
[OK] adceq pc, r0, r1
[KO] adcne pc, r0, r1
	pc: 0x300002 (UC) != 0x300000 (TT)
  c: 00000000 | 00000000 == 00000000
  n: 00000000 | 00000000 == 00000000
 pc: 00100004 | 00300002 != 00300000
 r0: 00100001 | 00100001 == 00100001
 r1: 00200000 | 00200000 == 00200000
r10: aca4baf2 | aca4baf2 == aca4baf2
r11: 4b711210 | 4b711210 == 4b711210
r12: 9bfd7b4c | 9bfd7b4c == 9bfd7b4c
r14: 1e747476 | 1e747476 == 1e747476
 r2: 00400001 | 00400001 == 00400001
 r3: 00100000 | 00100000 == 00100000
 r4: 00300001 | 00300001 == 00300001
 r5: 00600014 | 00600014 == 00600014
 r6: c09c9f2b | c09c9f2b == c09c9f2b
 r7: 11311fc3 | 11311fc3 == 11311fc3
 r8: 5bb0ce8f | 5bb0ce8f == 5bb0ce8f
 r9: ce2b72e0 | ce2b72e0 == ce2b72e0
 sp: 00500000 | 00500000 == 00500000
  v: 00000000 | 00000000 == 00000000
  z: 00000000 | 00000000 == 00000000

      Start 13: UnicornARM32Semantics5
13/41 Test #13: UnicornARM32Semantics5 ...........   Passed    0.23 sec
      Start 14: UnicornARM32Semantics6
14/41 Test #14: UnicornARM32Semantics6 ...........***Failed    0.18 sec
[KO] b #0x200000
	pc: 0x300002 (UC) != 0x300000 (TT)
  c: 00000001 | 00000001 == 00000001
  n: 00000000 | 00000000 == 00000000
 pc: 00100000 | 00300002 != 00300000
 r0: 00200001 | 00200001 == 00200001
 r1: adfa2575 | adfa2575 == adfa2575
r10: fbf2f109 | fbf2f109 == fbf2f109
r11: 5e4ec3de | 5e4ec3de == 5e4ec3de
r12: 5b457e4b | 5b457e4b == 5b457e4b
r14: 7ae6e12f | 7ae6e12f == 7ae6e12f
 r2: a1c7be69 | a1c7be69 == a1c7be69
 r3: 481b70d7 | 481b70d7 == 481b70d7
 r4: 2236b27e | 2236b27e == 2236b27e
 r5: 2166e238 | 2166e238 == 2166e238
 r6: 27ecad73 | 27ecad73 == 27ecad73
 r7: aa867667 | aa867667 == aa867667
 r8: 45646b93 | 45646b93 == 45646b93
 r9: cfdb02fc | cfdb02fc == cfdb02fc
 sp: 00500000 | 00500000 == 00500000
  v: 00000001 | 00000001 == 00000001
  z: 00000001 | 00000001 == 00000001

      Start 15: UnicornARM32Semantics7
15/41 Test #15: UnicornARM32Semantics7 ...........   Passed   47.97 sec
      Start 16: UnicornARM32Semantics8
16/41 Test #16: UnicornARM32Semantics8 ...........   Passed    0.54 sec
      Start 17: UnicornARM32Semantics9
17/41 Test #17: UnicornARM32Semantics9 ...........   Passed    0.16 sec
      Start 18: UnicornARM32Semantics10
18/41 Test #18: UnicornARM32Semantics10 ..........   Passed    0.21 sec
      Start 19: UnicornARM32Semantics11
19/41 Test #19: UnicornARM32Semantics11 ..........   Passed    0.16 sec
      Start 20: UnicornARM32Semantics12
20/41 Test #20: UnicornARM32Semantics12 ..........   Passed    0.78 sec
      Start 21: UnicornARM32Semantics13
21/41 Test #21: UnicornARM32Semantics13 ..........   Passed    7.72 sec
      Start 22: UnicornARM32Semantics14
22/41 Test #22: UnicornARM32Semantics14 ..........   Passed    0.22 sec
      Start 23: UnicornARM32Semantics15
23/41 Test #23: UnicornARM32Semantics15 ..........   Passed    0.22 sec
      Start 24: UnicornARM32Semantics16
24/41 Test #24: UnicornARM32Semantics16 ..........   Passed    0.17 sec
      Start 25: UnicornARM32Semantics17
25/41 Test #25: UnicornARM32Semantics17 ..........   Passed    0.22 sec
      Start 26: UnicornARM32Semantics18
26/41 Test #26: UnicornARM32Semantics18 ..........   Passed    0.33 sec
      Start 27: UnicornARM32Semantics19
27/41 Test #27: UnicornARM32Semantics19 ..........   Passed    0.17 sec
      Start 28: UnicornARM32Semantics20
28/41 Test #28: UnicornARM32Semantics20 ..........   Passed    0.17 sec
      Start 29: UnicornARM32Semantics21
29/41 Test #29: UnicornARM32Semantics21 ..........   Passed    0.18 sec
      Start 30: UnicornARM32Semantics22
30/41 Test #30: UnicornARM32Semantics22 ..........   Passed    1.48 sec
      Start 31: UnicornARM32Semantics23
31/41 Test #31: UnicornARM32Semantics23 ..........   Passed    0.31 sec
      Start 32: UnicornARM32Semantics24
32/41 Test #32: UnicornARM32Semantics24 ..........   Passed    0.24 sec
      Start 33: UnicornARM32Semantics25
33/41 Test #33: UnicornARM32Semantics25 ..........   Passed    0.27 sec
      Start 34: UnicornARM32Semantics26
34/41 Test #34: UnicornARM32Semantics26 ..........   Passed    0.28 sec
      Start 35: UnicornARM32Semantics27
35/41 Test #35: UnicornARM32Semantics27 ..........   Passed    0.25 sec
      Start 36: UnicornARM32Semantics28
36/41 Test #36: UnicornARM32Semantics28 ..........   Passed    1.39 sec
      Start 37: UnicornARM32Semantics29
37/41 Test #37: UnicornARM32Semantics29 ..........   Passed    0.32 sec
      Start 38: UnicornARM32Semantics30
38/41 Test #38: UnicornARM32Semantics30 ..........   Passed    0.27 sec
      Start 39: UnicornARM32Semantics31
39/41 Test #39: UnicornARM32Semantics31 ..........   Passed    0.34 sec
      Start 40: UnicornARM32Semantics32
40/41 Test #40: UnicornARM32Semantics32 ..........   Passed    0.36 sec
      Start 41: UnicornARM32Semantics33
41/41 Test #41: UnicornARM32Semantics33 ..........   Passed    0.34 sec

95% tests passed, 2 tests failed out of 41

Total Test time (real) =  75.57 sec

The following tests FAILED:
	 12 - UnicornARM32Semantics4 (Failed)
	 14 - UnicornARM32Semantics6 (Failed)
```

**Comment by cnheitman** [comment]
Yes, I'll take a look (this is weird, we did not change anything about this...)

**Comment by cnheitman** [comment]
Hey @JonathanSalwan !

Busy week last week... Anyways, I took a look at this.

Unicorn is behaving oddly when you add a `nop` after the branch (it executes it, despite I tell it to run only the branch instruction). This does not happen with version `1.0.2`.

The fix was easy (fb5e7b514afa6253454f66da28b69032473739fc), I just replaced the `nop` for another instruction. Now it works for both Unicorn versions. (The odd thing is that the instruction placed as the target of the branch is not even used, it's there in case we want to test more complex scenarios. I kept the code but changed the instruction.)
